packages/pds/src/api/com/atproto/repo/importRepo.ts

```diff
@@ -0,0 +1,128 @@
+import { Server } from '../../../../lexicon'
+import AppContext from '../../../../context'
+import { ActorStoreTransactor } from '../../../../actor-store'
+import { TID } from '@atproto/common'
+import {
+  Repo,
+  WriteOpAction,
+  getAndParseRecord,
+  readCarStream,
+  verifyDiff,
+} from '@atproto/repo'
+import { InvalidRequestError } from '@atproto/xrpc-server'
+import { CID } from 'multiformats/cid'
+import PQueue from 'p-queue'
+import { AtUri } from '@atproto/syntax'
+import { BlobRef, LexValue, RepoRecord } from '@atproto/lexicon'
+
+export default function (server: Server, ctx: AppContext) {
+  server.com.atproto.repo.importRepo({
+    auth: ctx.authVerifier.accessNotAppPassword,
+    handler: async ({ input, auth }) => {
+      const did = auth.credentials.did
+      if (!ctx.cfg.service.acceptingImports) {
+        throw new InvalidRequestError('Service is not accepting repo imports')
+      }
+      await ctx.actorStore.transact(did, (store) =>
+        importRepo(store, input.body),
+      )
+    },
+  })
+}
+
+const importRepo = async (
+  actorStore: ActorStoreTransactor,
+  incomingCar: AsyncIterable<Uint8Array>,
+) => {
+  const now = new Date().toISOString()
+  const rev = TID.nextStr()
+  const did = actorStore.repo.did
+
+  const { roots, blocks } = await readCarStream(incomingCar)
+  if (roots.length !== 1) {
+    throw new InvalidRequestError('expected one root')
```

Comment:
In the future we might look into failing earlier for this case which could feasibly be detected up front, just a slight vector for abuse.

Reply:
yeah that's a good point 👍 

we need a better CAR parsing lib first 😅